Many gcs

PalumboN opened the pull request:
_Done in DojoVM - Douai edition (and more)!_

Merged changes in https://github.com/pharo-project/pharo-vm/pull/649 (I cannot point to it because it comes from a **fork**, even though we work in the same team 😞)

So, this PR includes:

- Support for the new parameter in the CLI -> sets the GC at startup
- Transform polymorphic calls into switch-case

PalumboN committed:
Generating all GC strategies for the VM

Commit:
Added VMParameter for selecting garbage collector

Co-authored-by: Guille Polito <guillermopolito@gmail.com>
Co-authored-by: Nahuel Palumbo <nahuel.palumbo@gmail.com>

Commit:
added VM side code

Commit:
Passing GC parameter to the VM

Commit:
Passing GC parameter to the VM

PalumboN committed:
Select GC compactor at VM generation time

Commit:
Added support for fetching GC type from image

PalumboN committed:
Merge f10b7f761afdf7a57b2cdf8d13372d326ca58bd1

PalumboN committed:
Merge with terminal parameter

Commit:
Merge remote-tracking branch 'origin/dojo-douai' into dojo-douai

PalumboN committed:
Ups fixing merge

Commit:
Merge branch 'pharo-12' into dojo-douai

# Conflicts:
#	smalltalksrc/VMMaker/SpurMemoryManager.class.st
#	smalltalksrc/VMMaker/StackInterpreterPrimitives.class.st
#	src/parameters/parameters.c

Commit:
fixed merge again

Commit:
added API flag

PalumboN committed:
Slang: resolving polymorphic calls into switch-case

PalumboN committed:
Removing explicit sweepCompactor calls in MemoryManager code

guillep committed:
Merge branch 'pharo-12' into many-gcs

guillep commented on include/pharovm/parameters/parameters.h:
```diff
@@ -88,6 +88,13 @@ extern "C" {
+    HYBRID_COMPACTOR = 1,
+    SWEEP_COMPACTOR = 2,
+    SELECTIVE_COMPACTOR = 3
+} GarbageCollector;
```
Ok so here we have some indices,

guillep commented on smalltalksrc/Slang/CCodeGenerator.class.st:
```diff
@@ -481,6 +481,20 @@ CCodeGenerator >> arrayInitializerCalled: varName for: array sizeString: sizeStr
+		  newWithExpression: (TVariableNode named: variableToTest)
+		  selectors: (possibleClasses collect: [ :class |
+				   class staticallyResolvePolymorphicSelector: aSendNode selector ])
+		  arguments: aSendNode arguments
```
:)

guillep commented on smalltalksrc/Slang/CCodeGenerator.class.st:
```diff
@@ -4453,6 +4477,14 @@ CCodeGenerator >> reservedWords [
+
+	self staticallyResolvedPolymorphicReceivers
+		at: polymorphicVariable
+		put: { typeVariable } , classes
```
Maybe in a following iteration, we should have something better than an array here?

guillep commented on smalltalksrc/VMMaker/SpurCompactor.class.st:
```diff
@@ -31,7 +31,7 @@ Class {
 		'VMSpurObjectRepresentationConstants'
 	],
-	#category : #'VMMaker-SpurMemoryManager'
+	#category : #'VMMaker-InterpreterSimulation'
```
Why moved to simulation?

guillep commented on smalltalksrc/VMMaker/SpurCompactor.class.st:
```diff
@@ -46,13 +46,12 @@ SpurCompactor class >> declareCVarsIn: aCCodeGenerator [
 					staticallyResolveMethodNamed: selectorToStaticallyResolve
 					forClass: self
-					to: (self staticallyResolvePolymorphicSelector: selectorToStaticallyResolve)]]]
+					to: (self staticallyResolvePolymorphicSelector: selectorToStaticallyResolve)]]
```
Is this still needed?

guillep commented on smalltalksrc/VMMaker/SpurCompactor.class.st:
```diff
@@ -118,7 +110,7 @@ SpurCompactor >> scavenger: aSpurGenerationScavenger [
 
 { #category : #api }
-SpurCompactor >> shouldRemapObj: objOop [
+SpurCompactor >> shouldRemapObjCompactor: objOop [
```
For a next iteration, this should be done automatically :)

guillep commented on smalltalksrc/VMMaker/SpurMemoryManager.class.st:
```diff
@@ -713,35 +722,56 @@ SpurMemoryManager class >> compactorClass [
+				SpurPlanningCompactor.
+				SpurSweeper.
+				"SpurSelectiveCompactor" }.
+
```
Where is the mapping to what is specified in the command line?

    PLANNING_COMPACTOR = 0,
    HYBRID_COMPACTOR = 1,
    SWEEP_COMPACTOR = 2,
    SELECTIVE_COMPACTOR = 3

? Does this mean this PR should not be merged yet?

guillep commented on smalltalksrc/VMMaker/SpurMemoryManager.class.st:
```diff
@@ -713,35 +722,56 @@ SpurMemoryManager class >> compactorClass [
+
+	aCCodeGenerator
+		staticallyResolvedPolymorphicReceiver: 'planningCompactor'
+		to: SpurPlanningCompactor
```
Do we still need a planningCompactor here?

guillep commented on smalltalksrc/VMMaker/SpurMemoryManager.class.st:
```diff
@@ -948,7 +981,7 @@ SpurMemoryManager class >> isImmediate: anObject [
 { #category : #translation }
 SpurMemoryManager class >> isNonArgumentImplicitReceiverVariableName: aString [
-	^#('self' 'coInterpreter' 'manager' 'scavenger' 'segmentManager' 'compactor' 'planningCompactor' 'selectiveCompactor' 'heapMap') includes: aString
+	^#('self' 'coInterpreter' 'manager' 'scavenger' 'segmentManager' 'compactor' 'planningCompactor' 'planningCompactor' 'sweepCompactor' 'selectiveCompactor' 'heapMap') includes: aString
```
shouldn't we keep `compactor` only now?

guillep commented on smalltalksrc/VMMaker/SpurMemoryManager.class.st:
```diff
@@ -1961,6 +1994,10 @@ SpurMemoryManager >> allocateMemoryOfSize: memoryBytes newSpaceSize: newSpaceByt
 		             manager: self;
 		             yourself.
+		
+	planningCompactor := SpurPlanningCompactor simulatorClass new
```
and this too removed?

guillep commented on smalltalksrc/VMMaker/SpurMemoryManager.class.st:
```diff
@@ -6760,7 +6801,13 @@ SpurMemoryManager >> initializeObjectMemory: bytesToShift [
 	self updateFreeLists.
 	self computeFreeSpacePostSwizzle.
-	compactor postSwizzleAction.
+	
+"	gcType caseOf: { 
+		[ 0 ] -> [ planningCompactor postSwizzleAction ].
+		[ 1 ] -> [ sweepCompactor postSwizzleAction ].
+		ONLY SELECTIVE
+	 }.
+"
```
```suggestion
	compactor postSwizzleAction.
```

?

guillep commented on smalltalksrc/VMMaker/SpurMemoryManager.class.st:
```diff
@@ -11217,10 +11259,16 @@ SpurMemoryManager >> relocateObjStackForPlanningCompactor: objStack [
 					from: ObjStackFreex
-					to: ObjStackNextx + (self rawHashBitsOf: stackOrNil).
+					to: ObjStackNextx + (self rawHashBitsOf: stackOrNil) ].
+	 }.
```
Ok, this one needs polymorphism :)

guillep replied:
Ah, actually not, if we are here we know the compactor is always a planning compactor. Maybe we need to move this method to the compactor

guillep commented on smalltalksrc/VMMaker/SpurMemoryManager.class.st:
```diff
@@ -11229,10 +11277,14 @@ SpurMemoryManager >> relocateObjStackForPlanningCompactor: objStack [
 			from: ObjStackFreex
-			to: ObjStackFreex.
+			to: ObjStackFreex ].
+	 }.
```
same

guillep commented on src/client.c:
```diff
@@ -69,6 +70,7 @@ EXPORT(int) vm_init(VMParameters* parameters)
 	setMaxOldSpaceSize(parameters->maxOldSpaceSize);
 	setDesiredEdenBytes(parameters->edenSize);
 	setMinimalPermSpaceSize(parameters->minPermSpaceSize);
+    setGarbageCollectorType(parameters->garbageCollector);
```
```suggestion
	setGarbageCollectorType(parameters->garbageCollector);
```

guillep commented on src/parameters/parameters.c:
```diff
@@ -674,57 +659,58 @@ vm_parameters_parse(int argc, const char** argv, VMParameters* parameters)
+    parameters->garbageCollector = PLANNING_COMPACTOR;
+
+    return VM_SUCCESS;
+}
```
All the changes in this file are hidden in the re-formatting. Even hiding white spaces in the PR shows a lot of noise... It would have been nice to not have it...

PalumboN committed:
Merge branch 'dojo-douai' of github.com:BastouP411/pharo-vm into many-gcs

PalumboN committed:
Merge branch 'many-gcs' of github.com:PalumboN/opensmalltalk-vm into many-gcs

PalumboN committed:
Fixing Sweeper Compactor - First approach
(There is still a problem to transform different methods with the same selector)

guillep commented:
Migrated to https://github.com/evref-inria/pharo-vm/pull/10
